### PR TITLE
fix(labels): enforce per-tenant uniqueness on NormalizedName

### LINKED ISSUE #8090: Label.NormalizedName is unused by stores and has no per-tenant uniqueness
## Summary

`Label` maintains `NormalizedName` (`Name` setter lowercases into it) and the Labels API exposes it on create/update/get responses. Persistence does **not** use it:

- EF `Configurations` for `Label` is empty — no index, no unique constraint (contrast `WorkflowDefinitionLabel`, which at least has lookup indexes).
- Sqlite Labels snapshot: `Labels` table has only PK on `Id`; `NormalizedName` is a required column with no index.
- `ILabelStore` has no find-by-normalized-name; `ListAsync` / `FindById*` never touch `NormalizedName`.
- Create/Update endpoints do not reject duplicate names.

Meanwhile Secrets already enforces **per-tenant** uniqueness on normalized name (`IX_Secret_TenantId_NormalizedName`) with an explicit comment that a global unique would wrongly share names across tenants. Labels look like an unfinished copy of that pattern.

This is accidental complexity: a derived field + API surface with no store contract behind it. Both InMemory and EF are equally incomplete — not a Memory-only bug.

## Evidence

- `Label.Name` setter → `NormalizedName = value.ToLowerInvariant()`
- `Elsa.Persistence.EFCore/Modules/Labels/Configurations.cs` → `Configure(Label)` empty body
- Labels EF snapshot: no `NormalizedName` / `TenantId` indexes on `Labels`
- Secrets: `SecretConfiguration` unique index on `(TenantId, NormalizedName)`
- Create mapper trims `Name` and saves; no duplicate check

## Why it matters

- Operators can create duplicate label names (same tenant or cross-tenant) with different IDs; Studio/API consumers that key on name are ambiguous.
- `NormalizedName` in API responses implies a stable lookup key that the store does not honor.
- Subtractive clarity: either finish the contract (unique per tenant + optional find) or remove/stop exposing `NormalizedName` until needed.

## Proposed direction

Prefer one coherent choice:

1. **Finish (preferred if labels are name-keyed in UX):** add EF unique index on `(TenantId, NormalizedName)` (null-tenant policy consistent with other modules), enforce the same invariant in `InMemoryLabelStore.Save*`, reject duplicates at the API or store with a clear conflict, and add `FindByNormalizedName` only if callers need it.
2. **Subtract:** stop exposing `NormalizedName` on Labels DTOs and drop the derived field (or keep private) until a real lookup requirement appears — avoids half-implemented Secrets-shaped surface.

Do not add a second naming model. Align with Secrets’ per-tenant uniqueness rationale if finishing.

Also consider a non-unique `(WorkflowDefinitionVersionId, LabelId)` unique index on associations if Replace/Save paths can still insert duplicates outside the Diff-based Update endpoint — optional follow-on, not required for this issue’s core.

## Out of scope

- Labels Memory tenant isolation (separate issue).
- Secrets module changes.
- Studio labels UI (#3412).

## Milestone

Unset — Issue Triage / Crew Lead decide placement. Soft proposal: **3.9 Backlog**. Schema change if finishing uniqueness → coordinate with persistence owners (`schema change` label only if Triage agrees).

## Refs

- `src/modules/Elsa.Labels/Entities/Label.cs`
- `src/modules/Elsa.Persistence.EFCore/Modules/Labels/Configurations.cs`
- `src/modules/Elsa.Persistence.EFCore.Sqlite/Migrations/Labels/LabelsElsaDbContextModelSnapshot.cs`
- `src/modules/Elsa.Labels/Endpoints/Labels/Post/Mappers.cs`
- `src/modules/Elsa.Secrets.Persistence.EFCore/SecretConfiguration.cs` (finished pattern)
- Related theme: Labels Memory vs EF store-contract parity

### PR DESCRIPTION
**Pull request by sfmskywalker**
## Purpose

Make `Label.NormalizedName` a real store contract: unique per tenant in EF and Memory, matching Secrets' `IX_Secret_TenantId_NormalizedName`.

Fixes #8090.

---

## Scope

Select one primary concern:

- [x] Bug fix (behavior change)
- [ ] Refactor (no behavior change)
- [ ] Documentation update
- [ ] Formatting / code cleanup
- [ ] Dependency / build update
- [ ] New feature

---

### Problem

`Label.Name` already derives `NormalizedName`, and the Labels API returns it, but persistence never used it: no unique index, no Memory check, and create/update accepted duplicate names in the same tenant.

### Solution

- EF: unique index `IX_Label_TenantId_NormalizedName` on `(TenantId, NormalizedName)`. SQL Server/Oracle **keep** their filtered `TenantId IS NOT NULL` indexes. No `COALESCE` key and no second index.
- Memory: fail closed on `SaveAsync` / `SaveManyAsync`; clone-on-read so a rejected rename cannot mutate the stored row. Memory uniqueness semantics are unchanged (no null-vs-`""` remap).
- Default tenant is `""`, not null. Every provider migration runs `UPDATE Labels SET TenantId = '' WHERE TenantId IS NULL` before `CreateIndex`.
- Duplicate existing names are **not** auto-deleted (associations would be at risk). After the stamp, each provider **preflights leftover `(TenantId, NormalizedName)` keys (count + throw listing the keys)** and then `CreateIndex`es fail-loud if any remain.
- Shared `Label.NameMaxLength = 255` on `Name` and `NormalizedName` in shared Configurations (`HasMaxLength`). PostgreSQL/SQLite **keep provider column types** (`text` / `TEXT`). SQL Server/MySQL/Oracle narrow to 255; they **do not truncate**. Those providers **preflight over-length rows (count + throw listing Ids)** before `AlterColumn`.

---

## Upgrade notes

Operators must resolve leftover data **before** applying `PerTenantLabelUniqueness`. The migration will not delete or truncate rows.

1. After null `TenantId` is stamped to `''`, resolve any leftover rows that share `(TenantId, NormalizedName)`. The preflight lists the leftover keys and aborts; `CreateIndex` also fails loudly if any remain.
2. On SQL Server, MySQL, and Oracle: resolve any `Name` / `NormalizedName` longer than 255 characters. The preflight lists over-length Ids and aborts; `AlterColumn` is fail-closed and will not truncate.

---

## Verification

Local results at tip `7bff90e87`:

- `Elsa.Labels.UnitTests`: 31 passed
- Label-related `Elsa.Persistence.EFCore.UnitTests`: 19 passed

---

## Checklist

- [x] The PR is focused on a single concern
- [x] Commit messages follow the recommended convention
- [x] Tests added or updated (if applicable)
- [ ] Documentation updated (if applicable)
- [x] No unrelated cleanup included
- [x] All tests pass

<div><a href="https://cursor.com/agents/bc-6b990eab-cb39-5bd5-8e24-50a5bca3b71c?cursor_ref=pr_footer&cursor_cta=open_in_web"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-6b990eab-cb39-5bd5-8e24-50a5bca3b71c&cursor_ref=pr_footer&cursor_cta=open_in_cursor"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Comment by sfmskywalker** [comment]
@greptileai

**Review by sfmskywalker — COMMENTED** [review]
## Round 1 — REQUEST_CHANGES (HIGH)
**Tip:** `c3c05043ded88e366021c30bdbbeed2205bdfaf1` → `main` (#8090)

**Code Review gate:** REQUEST_CHANGES + HIGH  
**Greptile:** pending (dual gates still apply after fix).

### What looks good
- EF unique index `IX_Label_TenantId_NormalizedName` matches Secrets / Identity per-tenant shape; SQL Server/Oracle filtered unique on non-null `TenantId` matches Identity migrations.
- Memory fail-closed under `Sync` with same-Id rename, batch intra-collision, and batch swap (exclude batch Ids from store check) — clear and correct.
- `SyncNormalizedName` + dropping Update mapper overwrite that lowercased the untrimmed request name — right fix.
- Tests cover collide / rename-self / cross-tenant / ambient stamp / `*` / stale NormalizedName / SaveMany store+batch / swap.

### Blocking — same class as #8108 Greptile P1
`FindByIdAsync` / `FindManyByIdAsync` / `ListAsync` still hand out live `MemoryStore` refs.

Production path: `Labels.Update` does `FindById` → `UpdateEntity` (`e.Name = …`) → `SaveAsync`. `Label.Name` setter writes `NormalizedName` on the **live store object before** uniqueness runs. On duplicate, `SaveAsync` throws but the dictionary row is already renamed — and a second label can keep the old name, so Memory ends with two rows sharing a NormalizedName after a rejected update.

**Required (match #8108):**
1. Clone on `FindByIdAsync`, `FindManyByIdAsync`, and `ListAsync` page items (shallow field copy is fine).
2. Regression test: Find → rename onto collision → throw → stored Name/NormalizedName unchanged.

### Non-blocking
- SQL Server/Oracle filtered unique allows multiple null-`TenantId` rows with the same NormalizedName; Memory treats nulls as one tenant. Same Identity trade-off — OK.
- No `FindByNormalizedName` — correctly out of scope per issue.

Re-request Round 2 when the clone + test land.

**Comment by sfmskywalker** [comment]
@greptileai tip `4abb472` — Memory label Find/List return clones so a rejected rename cannot persist. Please re-review.

**Comment by sfmskywalker** [comment]
**Fixed in `4abb472`:** Memory label Find/List return clones so Labels.Update cannot mutate the stored row before Save; rejected duplicate leaves name/NormalizedName unchanged. Regression test added. Round 2 CR requested.

**Review by sfmskywalker — COMMENTED** [review]
## Round 2 — APPROVE (HIGH)
**Tip:** `4abb472da544adeb551b8fc6fd82a8affaf6453d` → `main` (#8090)  
Prior Round 1 REQUEST_CHANGES @ `c3c0504` is **STALE**.

**Code Review gate:** APPROVE + HIGH  
**Greptile:** pending 5/5 — dual gates required before merge.

### Delta
Closes the live-ref hazard from Round 1 (#8108 pattern):
- `FindByIdAsync` / `FindManyByIdAsync` / `ListAsync` return clones
- Explicit `Clone` with clear Find→mutate→Save comment
- Regression: Find → rename onto collision → throw → stored Name/NormalizedName unchanged

Uniqueness work from Round 1 still stands. Ship once Greptile hits 5/5.

**Comment by sfmskywalker** [comment]
@greptileai

**Review by sfmskywalker — COMMENTED** [review]
## Round 3 — APPROVE (HIGH)
**Tip:** `9613552d3ab32a76e98effad3309435e7554c453` → `main` (#8090)  
Prior Round 2 APPROVE @ `4abb472` is **STALE**.

**Code Review gate:** APPROVE + HIGH  
**Greptile:** was 2/5 on prior tip — awaiting re-score to 5/5. Dual gates still required. Note: Greptile’s “auto-reconcile/delete duplicates” P1 conflicts with Architect steer (fail-loud, no silent delete); CreateIndex failure is the intended loud stop.

### Delta vs Round 2
- All five providers: `UPDATE Labels SET TenantId='' WHERE TenantId IS NULL` before `CreateIndex`; SqlServer/Oracle keep filtered unique (`TenantId IS NOT NULL`) so stamped `""` rows are covered.
- No `DELETE` / `DeleteData` — duplicates fail at `CreateIndex` (migration test pins this).
- `Label.NameMaxLength = 255` + Name setter throw + EF `HasMaxLength` + AlterColumn (no silent truncate). Sqlite relies on app max-length (TEXT) — OK.
- Memory `UniquenessTenantId`: null ≡ `DefaultTenantId` (`""`), with test.
- Round 2 clone-on-read still present.

### Non-blocking
- Fail-loud is via `CreateIndex` rather than an explicit preflight COUNT/THROW — same outcome; clearer SQL message optional.
- Operator release note for pre-existing dups / >255 names still worth calling out in the PR/changelog.

Ship once Greptile hits 5/5.

**Comment by sfmskywalker** [comment]
@greptileai

**Comment by sfmskywalker** [comment]
@greptileai

**Review by sfmskywalker — COMMENTED** [review]
## Round 4/4 — APPROVE (HIGH)
**Tip:** `7bff90e870acde0cae311d074538c55df81c0bfb` → `main` (#8090)  
Prior Round 3 APPROVE @ `9613552` is **STALE**.  
(Requested tip was `672c721`; HEAD advanced with test-only `7bff90e` Oracle filter-escape assertion — included.)

**Code Review gate:** APPROVE + HIGH  
**Greptile:** pending 5/5 — dual gates. Greptile “auto-reconcile duplicates” remains intentionally declined (Architect: fail-loud).

### Delta
- Explicit fail-loud preflight listing leftover `(TenantId, NormalizedName)` keys before `CreateIndex` (all five providers: RAISE / THROW / SIGNAL / RAISE_APPLICATION_ERROR).
- Over-length Name/NormalizedName preflight before ALTER on SqlServer/MySQL/Oracle; PG/Sqlite keep unbounded provider types with model `HasMaxLength(255)` + entity setter (intentional).
- Round 3 null→`''` stamp, filtered SqlServer/Oracle unique, `NameMaxLength` 255, clone-on-read still present.
- Migration tests pin no DELETE, preflight keywords, filtered index, PG/Sqlite unbounded columns.

### Non-blocking
- Memory `UniquenessTenantId` (null≡`""`) from Round 3 was removed. Save still stamps via `ApplyCurrentTenant` before uniqueness — OK for the store API. Raw `MemoryStore` null leftovers would not collide with `""` (edge case only).

Final CR round complete. Ship once Greptile hits 5/5.

File: test/unit/Elsa.Persistence.EFCore.UnitTests/LabelPerTenantUniquenessMigrationTests.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public void PerTenantLabelUniqueness_KeepsFilteredUniqueIndex(string providerProject, string filter)
+    {
+        var migration = FindMigration(providerProject);
+        Assert.Contains(filter, migration, StringComparison.Ordinal);
+    }
+
+    private static string FindMigration(string providerProject)
+    {
+        var repoRoot = FindRepoRoot();
+        var labelsDir = Path.Combine(repoRoot, "src", "modules", providerProject, "Migrations", "Labels");
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/unit/Elsa.Persistence.EFCore.UnitTests/LabelPerTenantUniquenessMigrationTests.cs:82]: N/A for merge — test helper only (`FindMigration` path lookup). Segments are relative repo paths we control; no rooted segment risk in this fixture. Not product code; leaving Path.Combine as-is to avoid tip churn past Round 4 CR/Greptile gates.

File: test/unit/Elsa.Persistence.EFCore.UnitTests/LabelPerTenantUniquenessMigrationTests.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            .Single(file => !file.EndsWith(".Designer.cs", StringComparison.Ordinal));
+        return File.ReadAllText(path);
+    }
+
+    private static string FindRepoRoot()
+    {
+        var directory = new DirectoryInfo(AppContext.BaseDirectory);
+        while (directory is not null)
+        {
+            if (File.Exists(Path.Combine(directory.FullName, "Elsa.sln")))
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/unit/Elsa.Persistence.EFCore.UnitTests/LabelPerTenantUniquenessMigrationTests.cs:93]: N/A for merge — test helper only (`FindMigration` path lookup). Segments are relative repo paths we control; no rooted segment risk in this fixture. Not product code; leaving Path.Combine as-is to avoid tip churn past Round 4 CR/Greptile gates.
